pkgs/hooks_runner/lib/src/build_runner/build_planner.dart

```diff
@@ -126,40 +127,84 @@ class NativeAssetsBuildPlanner {
-      _timeAsync('BuildPlanner.makeBuildHookPlan', () async {
-        if (_buildHookPlan != null) return Success(_buildHookPlan!);
-        final packagesWithNativeAssets = await packagesWithHook(Hook.build);
-        if (packagesWithNativeAssets.isEmpty) {
-          // Avoid calculating the package graph if there are no hooks.
-          return const Success([]);
-        }
-        final packageMap = {
-          for (final package in packagesWithNativeAssets) package.name: package,
-        };
-        final packagesToBuild = packageMap.keys.toSet();
-        final stronglyConnectedComponents = packageGraph
-            .computeStrongComponents();
-        final result = <Package>[];
-        for (final stronglyConnectedComponent in stronglyConnectedComponents) {
-          final stronglyConnectedComponentWithNativeAssets = [
-            for (final packageName in stronglyConnectedComponent)
-              if (packagesToBuild.contains(packageName)) packageName,
-          ];
-          if (stronglyConnectedComponentWithNativeAssets.length > 1) {
-            logger.severe(
-              'Cyclic dependency for native asset builds in the following '
-              'packages: $stronglyConnectedComponentWithNativeAssets.',
-            );
-            return const Failure(HooksRunnerFailure.projectConfig);
-          } else if (stronglyConnectedComponentWithNativeAssets.length == 1) {
-            result.add(
-              packageMap[stronglyConnectedComponentWithNativeAssets.single]!,
-            );
-          }
-        }
-        _buildHookPlan = result;
-        return Success(result);
-      });
+      _timeAsync(
+        'BuildPlanner.makeBuildHookPlan',
+        () async => _makeHookPlan(
+          hookType: Hook.build,
+          cachedPlan: _buildHookPlan,
+          setCachedPlan: (plan) => _buildHookPlan = plan,
+          reverseOrder: false,
+        ),
+      );
+
+  /// Plans in what order to run link hooks.
+  ///
+  /// [PackageLayout.runPackageName] provides the entry-point in the graph. The
+  /// hooks of packages not in the transitive dependencies of
+  /// [PackageLayout.runPackageName] will not be run.
+  ///
+  /// Returns a [Future] that completes with a [Result]. On success, the
+  /// [Result] is a [Success] containing the [BuildPlan], which is a list of
+  /// packages in the order their link hooks should be executed. On failure, if
+  /// a cyclic dependency is detected among packages with native asset link
+  /// hooks, the [Result] is a [Failure] containing a
+  /// [HooksRunnerFailure.projectConfig].
+  Future<Result<BuildPlan, HooksRunnerFailure>> makeLinkHookPlan() async =>
+      _timeAsync(
+        'BuildPlanner.makeLinkHookPlan',
+        () async => _makeHookPlan(
+          hookType: Hook.link,
+          cachedPlan: _linkHookPlan,
+          setCachedPlan: (plan) => _linkHookPlan = plan,
+          reverseOrder: true, // Key difference here
```

Comment:
```suggestion
          reverseOrder: true, // Key difference from [makeBuildHookPlan]
```

(If you just land in this method by clicking through to it it will not be clear what the baseline of the comparison will be)